refactor(core): add static-query template strategy

devversion opened the pull request:
Introduces a new strategy for the `static-query` schematic that
is enabled by default. In order to provide a migration that works
for the most Angular applications and makes the upgrade as easy
as possible, the template strategy leverages the view engine
Angular compiler logic in order to determine the same query timing
that is currently used within applications using view engine.

devversion committed:
refactor(core): move usage detection into usage strategy

In order to support multiple strategies for detecting the query timing, the
query usage logic has been moved into a query usage strategy.

devversion committed:
refactor(compiler): export static-query detection logic

The `@angular/compiler` package currently contains the logic for determining whether
given queries are used statically or dynamically. This logic would be necessary in order
to build a schematic that leverages the Angular compiler API's in order to simulate the
query timing based on what ViewEngine computed at compilation-time/runtime.

Exporting the logic that is necessary to detect the timing should not affect the public
API as the `@angular/compiler` package is denoted as private in `PUBLIC_API.md`

alxhub reviewed (approved):
👍 for the compiler.

kara commented on packages/core/schematics/migrations/static-queries/transform.ts:
```diff
@@ -15,11 +15,24 @@ import {NgQueryDefinition, QueryTiming} from './angular/query-definition';
+                                     end: -1,
+                                     hasTrailingNewLine: false,
+                                     kind: ts.SyntaxKind.MultiLineCommentTrivia,
+                                     text: ' TODO '
```
Let's make this comment a bit more detailed, e.g.

```ts
'TODO: add missing static flag to query.'
```

devversion replied:
Can we make this decision dependent on the result of the comment below? Either way we can add such a message into the TODO comment (regardless of inline; or a line above)

As mentioned in the comment below: The idea was to print the messages _only_ in console to not pollute the user-code too much.

kara replied:
See comment in other thread

kara commented on packages/core/schematics/migrations/static-queries/transform.ts:
```diff
@@ -15,11 +15,24 @@ import {NgQueryDefinition, QueryTiming} from './angular/query-definition';
   const timingPropertyAssignment = ts.createPropertyAssignment(
-      'static', timing === QueryTiming.STATIC ? ts.createTrue() : ts.createFalse());
+      'static', timing === null ?
+          ts.createIdentifier('undefined') :
```
Rather than setting a flag to "undefined", let's just not set any flag and add a TODO in the line above.

devversion replied:
It's not that easy to insert a comment before the property declaration. This is because indentation can be different and we would need to use the `ts.Printer` to re-print the whole class then. This causes unnecessary formatting of the whole class (which is undesired).

Also the idea was to print the error message in the console and just add the TODO to imply where something needs to be changed (we also do inline-todo's in material)

kara replied:
If moving the comment is a pain, I'm fine with keeping the TODO where it is. That said, I don't think a comment that simply reads `TODO` is helpful to have in a codebase. It's just noise with no indication of what should change. Perhaps a shorter version:

```
TODO: add static flag
```

devversion replied:
Agreed that a bit context in the TODO is helpful. Updated it to say `TODO: add static flag`. Also no longer adding `undefined`, but rather just something like:

````ts
@ViewChild(selector, /* TODO: add static flag */ {})
```

kara commented on packages/core/schematics/test/static_queries_migration_template_spec.ts:
```diff
@@ -0,0 +1,485 @@
+          </ng-template>
+          <div>
+            <button #myStaticButton>Button outside ng-template</button>
+          </div>
```
Can we also add a test where we query for a local ref on a template element itself? eg.

```html
<ng-template #contentTemplate>
    ....
</ng-template>
```

A query for `'contentTemplate'` should have `static: true`.

devversion replied:
yep. good case. will add a test.

kara reviewed (changes requested):
Approach looks good, some minor changes requested for the printed comment / tests

devversion commented on packages/core/schematics/test/static_queries_migration_template_spec.ts:
```diff
@@ -0,0 +1,485 @@
+          .toContain(`@ViewChild('my-token-2', { static: false }) query2: any;`);
+    });
+
+    it('should detect queries selecting component view providers through string token', () => {
```
Good catch. always used the previous test case as boilerplate.

devversion commented on packages/core/schematics/test/static_queries_migration_template_spec.ts:
```diff
@@ -0,0 +1,485 @@
+      runMigration();
+
+      expect(tree.readContent('/index.ts'))
+          .toContain(`@ViewChild(HomeComponent, { static: true }) query: any;`);
```
yep. added.

devversion committed:
refactor(core): add static-query template strategy

Introduces a new strategy for the `static-query` schematic that
is enabled by default. In order to provide a migration that works
for the most Angular applications and makes the upgrade as easy
as possible, the template strategy leverages the view engine
Angular compiler logic in order to determine the query timing
that is currently used within applications using view engine.

kara reviewed (approved):
LGTM

kara commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:243290809:BASE:243290871:1555090996948:83336ef1)

kara commented:
caretaker note: will need a patch to angular BUILD to add dependency
